Specs/AESSpec.lean

```diff
@@ -0,0 +1,182 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Hanno Becker
+-/
+
+import Arm.BitVec
+import Arm.Vec
+
+import Lean
+open Lean Meta
+
+open BitVec
+
+/-
+  The goal of this file is to provide a model of the AES specification
+  that is as close as possible to the original FIPS document.
+
+  TODO: This is work in progress. For the moment, we only have definitions
+    of the AES state as a bitvector, byte vector, or byte array, as well as
+    conversions between them.
+-/
+
+/-
+   3.3 Indexing of Byte Sequences
+-/
+
+abbrev Byte : Type := BitVec 8
+
+def bitvec_split_byte {k : Nat} (invec : BitVec (8*k)) : Byte × (BitVec (8*(k-1))) :=
+  let byte : BitVec 8 := invec.truncate 8
+  let remainder := BitVec.extractLsb' 8 (8*(k-1)) invec
+  (byte, remainder)
+
+theorem bitvec_split_byte_append {k : Nat} (v : BitVec (8*(k-1))) (b : Byte) (h : 8 * (k-1) + 8 = 8 * k) :
+    bitvec_split_byte (BitVec.cast h (v ++ b)) = (b, v) := by simp [bitvec_split_byte]
+
+theorem bitvec_split_byte_append' {k : Nat} (v : BitVec (8*(k+1))):
+   let (b, v') := bitvec_split_byte v
+   v' ++ b = v := by
+  apply eq_of_getLsb_eq
+  intro i
+  simp [getLsb_append]
+  by_cases h: ((i : Nat) < 8)
+  · simp [h]
+  · simp [h]
+    have lt: (i : Nat) - 8 < 8*k := by simp_all; omega
+    have e : (8 + ((i : Nat) - 8)) = i := by simp_all
+    simp [e, lt]
+    done
+  done
+
+-- Splitting a bitvector of 8*k entries into bytes, little endian
+def bitvec_to_byte_seq (k : Nat) (invec: BitVec (8*k)) : Vec Byte k :=
+  if k_gt_0: (k > 0) then
+    let (byte, remainder) := bitvec_split_byte invec
+    have h: k - 1 + 1 = k := by omega
+    h ▸ Vec.cons byte (bitvec_to_byte_seq (k-1) remainder)
+ else
+    have h: k = 0 := by omega
+    h ▸ Vec.empty
+
+-- Concatenating a little endian byte sequence into a bitvector
+def byte_seq_to_bitvec: Vec Byte k → BitVec (8*k)
+  | ⟨[], h⟩ =>
+      let h' : 0 = 8*k := by simp_all; omega
+      BitVec.cast h' (BitVec.ofNat 0 0)
+  | ⟨a :: as, h⟩ =>
+      let g : 8 * (k - 1) + 8 = 8 * k := by simp [←h]; omega
+      BitVec.cast g (byte_seq_to_bitvec ⟨as, by simp[←h]⟩ ++ a)
+
+@[simp]
+theorem byte_seq_to_bitvec_cons:
+  byte_seq_to_bitvec (Vec.cons byte v) = (byte_seq_to_bitvec v) ++ byte := by
+    cases v; simp only [Vec.cons, byte_seq_to_bitvec]; rfl
+
+-- Example
+def example_bitvec : BitVec 128 := 0x0102030405060708090a0b0c0d0e0f#128
+def example_byteseq  := bitvec_to_byte_seq 16 example_bitvec
+def example_bitvec'  := byte_seq_to_bitvec example_byteseq
+def example_byteseq' := bitvec_to_byte_seq 16 example_bitvec'
+
+#eval example_bitvec
```

Comment:
I'd recommend using `example` to test these out instead of using `#eval`. E.g., see https://github.com/leanprover/LNSym/blob/bdfadd596f64940993650e6378543a9ddebd27a1/Arm/Decode.lean#L165.